edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs

```diff
@@ -238,9 +249,57 @@ async void OnConnectionStatusChanged(ConnectionStatus status, ConnectionStatusCh
 
                 if (this.pullOnReconnect && this.initTask.IsCompleted && status == ConnectionStatus.Connected)
                 {
+                    var delayedTwinPull = true;
                     using (await this.twinLock.LockAsync())
                     {
-                        await this.RefreshTwinAsync();
+                        var now = DateTime.Now;
+                        if (now - this.lastTwinPullOnConnect > this.TwinPullOnConnectThrottleTime && !this.isDelayedTwinPullInProgress.Get())
+                        {
+                            this.lastTwinPullOnConnect = now;
+                            await this.RefreshTwinAsync();
+                            delayedTwinPull = false;
+                        }
+                    }
+
+                    if (delayedTwinPull)
+                    {
+                        if (this.isDelayedTwinPullInProgress.GetAndSet(true))
+                        {
+                            this.pullRequestCounter++;
+                        }
+                        else
+                        {
+                            _ = Task.Run(
+                                async () =>
+                                {
+                                    await Task.Delay(this.TwinPullOnConnectThrottleTime);
+
+                                    var requestCounter = default(int);
+                                    using (await this.twinLock.LockAsync())
+                                    {
+                                        this.lastTwinPullOnConnect = DateTime.Now;
+
+                                        try
+                                        {
+                                            await this.RefreshTwinAsync();
+                                        }
+                                        catch
+                                        {
+                                            // swallowing intentionally
+                                        }
+
+                                        requestCounter = this.pullRequestCounter;
+                                        this.pullRequestCounter = 0;
+
+                                        this.isDelayedTwinPullInProgress.Set(false);
+                                    }
+
+                                    if (requestCounter > PullFrequencyThreshold)
+                                    {
+                                        Events.PullingTwinHasBeenTriggeredFrequently(requestCounter, Convert.ToInt32(this.TwinPullOnConnectThrottleTime.TotalSeconds));
```

Comment:
This is great.. 
Can you just add Debug logs for all the other events as well - that way we can trace how many times this is getting called, and the steps the code takes. 
(for example, what was the requestCounter, etc.)